File Browser Filter

echarles opened the pull request:
## References

This PR adresses https://github.com/jupyterlab/jupyterlab/issues/8157.

For now, only the files of the current folder are considered, see also the TODO section hereafter.

## Code changes

- Changes in `filebrowser` and `filebrowser-extension` TS packages.
- Use `FilterFileBrowserModel` instead of `FileBrowserModel` and create a `search.ts` for input text

## User-facing changes

![filefilter](https://user-images.githubusercontent.com/226720/85788354-1f9cc300-b72d-11ea-8374-14abd7c59e22.gif)

## Backwards-incompatible changes

None.

## TODO

- [ ] Fix CSS issue when file listing is larger than the display layout.
- [ ] Consider to filter on folders (the existing `FilterFileBrowserModel` keeps the folders.
- [ ] Consider to filter on subfolders if this feature is required.
- [ ] Consider highlighting in the filenames the given filter.

telamonian commented:
@echarles Looks good so far. I think the UI needs some work.

- insert the search bar above the breadcrumbs
  - the search bar probably shouldn't cover up the breadcrumbs, because then there's no info displayed to the user about what their cwd is. Instead, insert the search UI above the breadcrumb

- get rid of the search toolbar button
  - at the same time, I *hate* list UI's that shift large elements up and down depending on state. Having things jump up and down makes me seasick. I think most users would appreciate the existence of a search bar, so just get rid of the toolbar button and have it always visible
  - the issue could also be mitigated by animating the transition, but that's much fancier than our usual approach to styling

- Change the placeholder text to `filter files by name`
  - intuitively, I think most users would expect a "file search" functionality to recurse into dirs, or possibly search their whole tree starting from root
  - since the current proposal is for a shallow search, it would be clearer to call this a "file filter"

echarles commented:
Thx @telamonian for your review and feedbacks. I will push some changes to take them.

> at the same time, I hate list UI's that shift large elements up and down depending on state. Having things jump up and down makes me seasick. I think most users would appreciate the existence of a search bar, so just get rid of the toolbar button and have it always visible

I would like just to think a bit more to having the input search box always visible instead of activating it via the toolbar and replacing the breadcrumbs. With the `always visible` option, we loose some space on the file browser area. I am fine for that, but we may need to think a bit more and maybe have other opinions and options.

cc/ @tgeorgeux

echarles commented:
I have updated the placeholder which is now `Filter files by name` as @telamonian has suggested.

For the search box UI, I have also implemented @telamonian suggestion. I will put this topic at today weekly meeting to gather broader community feedback as this change is quite visible for the user. I see 3 options:

1. Toogle via toolbar (replace breadcrumbs)

<img width="294" alt="Screenshot 2020-07-01 at 15 25 52" src="https://user-images.githubusercontent.com/226720/86250888-741ec300-bbb1-11ea-8cba-7d135c588ecc.png">

2. Always visible

<img width="324" alt="Screenshot 2020-07-01 at 15 30 31"  #src="https://user-images.githubusercontent.com/226720/86250904-7c76fe00-bbb1-11ea-8188-715c6b0beb73.png">

3. Side-panel (a-la-vscode)

<img width="296" alt="Screenshot 2020-07-01 at 15 36 40" src="https://user-images.githubusercontent.com/226720/86250935-87ca2980-bbb1-11ea-8fdd-74f5e8ef7f60.png">

You can also mix them, e.g. 1. Toogle via toolbar + keep the breadcrumbs. Side-panel (a-la-vscode) would have more impact in terms of implementation, and also makes less sense for now as the goal of this PR is restricted to filter on files in the current directory, so not a real search on file names in the nested folders.

tgeorgeux commented:
Option 2 follows the design patterns the best at this point. I would make a few small adjustments:
- Add 4px padding all around the search box.
- Switch out the Icon for the [JupyterLab Search Icon ](https://github.com/jupyterlab/jupyterlab/blob/master/packages/ui-components/style/icons/toolbar/search.svg)(we are using material Icons, that one looks like it's from the Ionic Icon set).

echarles commented:
@tgeorgeux padding and icon updated.

@telamonian Tim has asked to use mui icon, so I have update the InputGroup rightIcon to use LabIcon (and impact the ones using that) - Can you have a look at that and provide any feedback (eg. I use `rightIcon="ui-components:search"`, not sure if you would prefer `rightIcon="search"`

<img width="219" alt="Screenshot 2020-07-01 at 17 48 21" src="https://user-images.githubusercontent.com/226720/86265431-10ea5c00-bbc4-11ea-8adc-8cf7602ed34f.png">

The impact of this is that eg. extension manager now also uses the LabIcon, which was not the case before (Discussed about that with Tim and he said he would prefer to align all search icons anyway).

<img width="214" alt="Screenshot 2020-07-01 at 17 48 33" src="https://user-images.githubusercontent.com/226720/86265546-34ada200-bbc4-11ea-914f-d48039a68b98.png">

echarles commented:
@telamonian I think you said filter *above* the breadcrumbs, so here it is

<img width="217" alt="Screenshot 2020-07-01 at 18 07 42" src="https://user-images.githubusercontent.com/226720/86266667-d386ce00-bbc5-11ea-8e6c-0e7d6d0839e0.png">

blink1073 commented:
Command palette uses [`matchSumOfDeltas`](https://github.com/jupyterlab/lumino/blob/52b15d88cd0e647c40bd1d6a265817b7cc6458a2/packages/algorithm/src/string.ts#L132) to do its [filtering](https://github.com/jupyterlab/lumino/blob/7831eb2725f291a6c99fc72ae0b9be42afe0ca8c/packages/widgets/src/commandpalette.ts#L1246).

echarles commented:
Thx @blink1073. Very useful links that helped me to go fuzzy (with a small optional addition attributed on the IContent model). This is ready for review.

![filefilter](https://user-images.githubusercontent.com/226720/86365816-4b163500-bc7a-11ea-866d-8cb347b74603.gif)

blink1073 commented on packages/filebrowser/src/search.tsx:
```diff
@@ -0,0 +1,137 @@
+};
+
+/**
+ * A widget which hosts a ...
```
Missing an ending ;)

blink1073 commented:
Let's make sure this still passes after merging #8666

blink1073 commented:
Kicking CI, I'll cut an alpha release after we merge this.

blink1073 commented:
Looks like the filebrowser example needs updating: 

```
##[error]src/index.ts(97,5): error TS2739: Type 'FileBrowserModel' is missing the following properties from type 'FilterFileBrowserModel': setFilter, _filter
info Visit https://yarnpkg.com/en/docs/cli/run for documentation about this command.
```

echarles commented:
@blink1073 Example is fixed and CI has been green

@telamonian I have added a user setting `useFuzzyFilter` which is true by default - Set it to false for normal filter

@tgeorgeux The mapped string are now shown like in the command sidebar (black bold)

echarles commented:
![ezgif-1-23da821276c0](https://user-images.githubusercontent.com/226720/87066311-a31ce080-c212-11ea-867a-91e3f3099402.gif)

blink1073 commented:
Testing the new auto-cancel bot for GH Actions, opening and closing a couple times.

blink1073 commented:
Success!

blink1073 commented:
Failure in js-filebrowser:

```
 FAIL  test/openfiledialog.spec.ts
  ● @jupyterlab/filebrowser › FileDialog.getExistingDirectory() › should return one selected directory

    expect(received).toBe(expected) // Object.is equality

    Expected: "directory"
    Received: "file"
```

echarles commented:
@blink1073 My last commit had unexpected side effect which I have fixed now. Thx for spotting this

CI is green (except Linux Tests / Linux (docs, 3.8) which fails after 78m, unrelated to this PR).

blink1073 commented:
Nice work @echarles!

michaelaye commented:
Is there a work-around to enable filtering on folders in JLab 3.0 ?

jpcaram commented:
I would like to filter folders too! Is there a setting?

vidartf commented:
> I would like to filter folders too! Is there a setting?

AFAIK, there is no such setting. You should probably make a new issue with this feature request if you want it 👍
